multy_core/src/bitcoin/bitcoin_transaction_segwit.cpp

```diff
@@ -0,0 +1,506 @@
+    }
+
+    return make_script_pub_key(*binary_address, address_type);
+}
+
+
+
+} // namespace
+
+namespace multy_core
+{
+namespace internal
+{
+
+BitcoinTransactionSegWitDestination::BitcoinTransactionSegWitDestination(BitcoinNetType net_type)
+    : amount(m_properties,
+             "amount",
+             Property::REQUIRED,
+             &verify_bigger_than<BigInt, 0, ERROR_INVALID_ARGUMENT>),
+      address(m_properties,
+              "address",
+              Property::REQUIRED,
+              [this](const std::string &new_address) {
+                    this->sig_script = make_script_pub_key_from_address(this->m_net_type, new_address);
+              }),
+      is_change(m_properties,
+                "is_change",
+                Property::OPTIONAL,
+                [](const int32_t value) {
+                    if (value != 0 )
+                    {
+                        THROW_EXCEPTION("NOT SET is_change to segwit TX");
+                    }
+                }),
+      sig_script(),
+      m_net_type(net_type)
+{
+}
+
+//BinaryData BitcoinTransactionSegWitDestination::preview_output()
+//{
+//    BitcoinDataStream result;
+
+//    result << amount;
+//    result << sig_script;
+
+//    return result.get_content();
+//}
+
+BitcoinTransactionSegWitSource::BitcoinTransactionSegWitSource()
+    : prev_transaction_hash(
+              m_properties,
+              "prev_tx_hash",
+              Property::REQUIRED,
+              [](const BinaryData& new_tx_out_hash) {
+                  if (new_tx_out_hash.len != 32)
+                  {
+                      THROW_EXCEPTION2(
+                              ERROR_TRANSACTION_SOURCE_INVALID_PREV_TX_HASH,
+                              "Previous transaction hash should be"
+                              "exactly 32 bytes long.");
+                  }
+              }),
+      prev_transaction_out_index(
+              m_properties,
+              "prev_tx_out_index",
+              Property::REQUIRED),
+      prev_transaction_out_script_pubkey(
+              m_properties,
+              "prev_tx_out_script_pubkey",
+              Property::REQUIRED),
+      private_key(
+              m_properties,
+              "private_key",
+              Property::REQUIRED),
+      sequence(make_clone(BinaryData{BITCOIN_INPUT_SEQ_FINAL, 4}),
+                m_properties, "sequence", Property::OPTIONAL),
+      script_signature(),
+      script_witness(),
+      amount(m_properties, "amount", Property::REQUIRED)
+{
+}
+
+void BitcoinTransactionSegWitSource::sign(hash<256>& hash_prevouts, hash<256>& hash_sequence, hash<256>& hash_outputs, BinaryDataPtr& lock_time)
+{
+
+    const PublicKeyPtr& public_key = private_key.get_value()->make_public_key();
+    std::array<uint8_t, HASH160_LEN> segwit_script_P2WPKH;
+    BinaryData public_key_hash_data = as_binary_data(segwit_script_P2WPKH);
+    bitcoin_hash_160(public_key->get_content(), &public_key_hash_data);
```

Comment:
This is used so often and looks so ugly, that it might be a good time to add a `BITCOIN_HASH` as supported hash for `do_hash()` function.

It would then become:
```
const auto segwit_script_P2WPKH = do_hash<BITCOIN_HASH, 160>(private_key->make_public_key()->get_content());
```